### PR TITLE
Add HashTree to the Prelude

### PR DESCRIPTION
**Pull request by iamrecursion**
### Pull Request Description
Adds the `HashTree` type to the prelude.

### Important Notes
It uses criterion for its benchmarks, but this doesn't preclude use of the standard harness elsewhere.

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guide.
- [x] All code has been tested where possible.

### CODE REVIEW

File: src/data/src/hash_map_tree.rs [hunk 1]
```diff
@@ -243,6 +263,51 @@ where K:Eq+Hash {
...
+    /// ## NOTE
+    /// This function is only suitable for use on trees with small depths as it is implemented in a
+    /// recursive fashion.
+    pub fn map_in_place<F>(&mut self, f:F)
+        where F : Copy + Fn(&mut V) -> () {
+        f(&mut self.value);
+        for value in self.branches.values_mut() {
+            value.map_in_place(f);
+        }
+    }
```
Comment (wdanilo) [inline src/data/src/hash_map_tree.rs:309]: do it in the rust way - like in Vec - implement `iter_mut` instead.

Comment (iamrecursion) [reply src/data/src/hash_map_tree.rs:309]: Turns out it already has a derived mutable iterator that I missed.

**Review by farmaazon — APPROVED** [review]

File: src/data/src/hash_map_tree.rs [hunk 2]
```diff
@@ -175,7 +210,7 @@ where K : Eq+Hash,
     /// Zips two trees together into a new tree with cloned values.
     #[inline]
     pub fn zip_clone<T2>
-    (&self, other:&HashMapTree<K,T2,S>) -> HashMapTree<K,AtLeastOneOfTwo<T,T2>,S>
+    (&self, other: &HashMapTree<K,T2,S>) -> HashMapTree<K,AtLeastOneOfTwo<T,T2>,S>
```
Comment (wdanilo) [inline src/data/src/hash_map_tree.rs:213]: spacing - old version was ok

File: src/data/src/hash_map_tree.rs [hunk 3]
```diff
@@ -235,7 +270,7 @@ where K:Eq+Hash {
 
     /// Gets the current value or creates new one if missing.
     pub fn value_or_set_with<F>(&mut self, cons:F) -> &mut T
-    where F:FnOnce()->T {
+        where F: FnOnce() -> T {
```
Comment (wdanilo) [inline src/data/src/hash_map_tree.rs:273]: indentation - old version was ok

File: src/data/src/hash_map_tree.rs [hunk 4]
```diff
@@ -247,22 +282,22 @@ where K:Eq+Hash {
...
 
 impl<K,V,S> PartialSemigroup<&HashMapTree<K,V,S>> for HashMapTree<K,V,S>
-    where K : Eq + Hash + Clone,
-          V : Semigroup,
-          S : BuildHasher + Clone {
-    fn concat_mut(&mut self, other:&Self) {
+where K : Eq+Hash+Clone,
+      V : Semigroup,
+      S : BuildHasher+Clone {
+    fn concat_mut(&mut self, other: &Self) {
```
Comment (wdanilo) [inline src/data/src/hash_map_tree.rs:298]: spacing, old version was ok

**Review by wdanilo — APPROVED** [review]

**Pull request merged**